core/shared/src/main/scala/zio/ZLayer.scala

```diff
@@ -474,6 +474,150 @@ object ZLayer extends ZLayerCompanionVersionSpecific {
   ): ZLayer[RIn, E, ROut] =
     ZLayer.fromZIO(zio)
 
+  /**
+   * Defines lifecycle methods to be utilized during the automatic derivation of
+   * `ZLayer` using [[ZLayer.derive]].
+   *
+   * This trait allows types to specify initialization and cleanup logic during
+   * their construction and destruction, respectively, when a `ZLayer` for the
+   * type is automatically derived. It's especially useful in scenarios where
+   * resources need to be set up or torn down as part of the derived layer's
+   * lifecycle.
+   *
+   * @note
+   *   This trait's lifecycle hooks are specifically designed to work with
+   *   [[ZLayer.derive]]. Using it outside this context won't inherently attach
+   *   any lifecycle behaviors to the type.
+   *
+   * Implementors should define the `initialize` and `cleanup` methods to
+   * provide the desired lifecycle behaviors.
+   */
+  trait LifecycleHooks[-R, +E] {
+    def initialize: ZIO[R, E, Any]
+    def cleanup: ZIO[R, Nothing, Any]
+  }
+
+  /**
+   * Provides a default way to construct or provide an instance of type `A`.
+   *
+   * Used during `ZLayer` derivation to resolve dependencies. If an implicit
+   * `Default[A]` instance exists for a type, it signifies that a default value
+   * can be used, bypassing the dependency in the `ZLayer` environment.
+   *
+   * @note
+   *   When explicitly type-annotating the implicit val, ensure it's in the form
+   *   `ZLayer.Default.Aux[R, E, A]` rather than just `ZLayer.Default[A]` to
+   *   ensure correct type inference and dependency resolution during `ZLayer`
+   *   derivation.
+   */
+  trait Default[+A] {
+    type R
+    type E
+
+    def layer: ZLayer[R, E, A]
+  }
+  object Default extends DefaultInstances0 {
+
+    type Aux[R0, E0, A] = Default[A] {
```

Comment:
Would be good to have a more domain specific name here.

Reply:
How about `Resolved`?

Reply:
`WithContext`?